runtime/parachains/src/inclusion/mod.rs

```diff
@@ -402,24 +406,24 @@ impl<T: Config> Pallet<T> {
 			}
 		}
 
-		freed_cores
+		(freed_cores, enacted_candidate_weight)
 	}
 
 	/// Process a set of incoming bitfields.
 	///
 	/// Returns a `Vec` of `CandidateHash`es and their respective `AvailabilityCore`s that became available,
-	/// and cores free.
+	/// and cores free. Additionally returns the weight consumed by enacted candidates.
 	pub(crate) fn process_bitfields(
 		expected_bits: usize,
 		signed_bitfields: UncheckedSignedAvailabilityBitfields,
 		disputed_bitfield: DisputedBitfield,
 		core_lookup: impl Fn(CoreIndex) -> Option<ParaId>,
-	) -> Vec<(CoreIndex, CandidateHash)> {
+	) -> (Vec<(CoreIndex, CandidateHash)>, Weight) {
```

Comment:
From a design perspective and writing test cases, it'd make more sense to return a `Vec<CandidateHash>`, and the caller can do the weight calc.

Reply:
So right now the path the weight gets bubbled up is `enact_candidate
 -> update_pending_availability_and_get_freed_cores -> process_bitfields`. In `enact_candidate` we go through and track the actual weights consumed (opposed to the worst case estimates we use prior).

With the current function having a seperate functionw would be tricky and require some duplicate logic. Basically we would need to duplicate the logic that gets us to here: https://github.com/paritytech/polkadot/blob/888162259348c7fa3e9ad7454e8056235565efa6/runtime/parachains/src/inclusion/mod.rs#L370-L402 . At the moment we are actually modifying storage right there, so we essentially would need get the weight first of enacting candidates without side effects and then call `update_pending_availability_and_get_freed_cores`, which would duplicate the filtering logic but actually enact the candidate and write to storage

Reply:
In an `enact_candidate_weight` function we would need to take the weight calculations done within these calls https://github.com/paritytech/polkadot/blob/888162259348c7fa3e9ad7454e8056235565efa6/runtime/parachains/src/inclusion/mod.rs#L748-L772 and abstract them to their own weight only functions

Reply:
While I see the general design merit of having the weight tracked by weight only functions - in this particular case there is a lot of branching logic to determine how much weight actually gets used and duplicating much of that logic in the weight only functions since harder to keep track of then tracking the weight as the actual execution takes place.

Reply:
I see that this is not feasible in the remaining time, could you create an issue outlining the merits and problems with such an approach?

Reply:
https://github.com/paritytech/polkadot/issues/4511